zcash_primitives/src/transaction/builder.rs

```diff
@@ -0,0 +1,692 @@
+
+        Ok(())
+    }
+
+    /// Adds a transparent address to send funds to.
+    pub fn add_transparent_output(
+        &mut self,
+        to: &TransparentAddress,
+        value: Amount,
+    ) -> Result<(), Error> {
+        if value.0 < 0 {
+            return Err(Error(ErrorKind::InvalidAmount));
+        }
+
+        self.mtx.vout.push(TxOut {
+            value,
+            script_pubkey: to.script(),
+        });
+
+        Ok(())
+    }
+
+    /// Sets the Sapling address to which any change will be sent.
+    ///
+    /// By default, change is sent to the Sapling address corresponding to the first note
+    /// being spent (i.e. the first call to [`Builder::add_sapling_spend`]).
+    pub fn send_change_to(&mut self, ovk: OutgoingViewingKey, to: PaymentAddress<Bls12>) {
+        self.change_address = Some((ovk, to));
+    }
+
+    /// Builds a transaction from the configured spends and outputs.
+    ///
+    /// Upon success, returns a tuple containing the final transaction, and the
+    /// [`TransactionMetadata`] generated during the build process.
+    ///
+    /// `consensus_branch_id` must be valid for the block height that this transaction is
+    /// targeting. An invalid `consensus_branch_id` will *not* result in an error from
+    /// this function, and instead will generate a transaction that will be rejected by
+    /// the network.
+    pub fn build(
+        mut self,
+        consensus_branch_id: u32,
+        prover: impl TxProver,
+    ) -> Result<(Transaction, TransactionMetadata), Error> {
+        let mut tx_metadata = TransactionMetadata::new();
+
+        //
+        // Consistency checks
+        //
+
+        // Valid change
+        let change = self.mtx.value_balance.0
+            - self.fee.0
+            - self
+                .mtx
+                .vout
+                .iter()
+                .map(|output| output.value.0)
+                .sum::<i64>();
+        if change.is_negative() {
+            return Err(Error(ErrorKind::ChangeIsNegative(change)));
+        }
+
+        //
+        // Change output
+        //
+
+        if change.is_positive() {
+            // Send change to the specified change address. If no change address
+            // was set, send change to the first Sapling address given as input.
+            let change_address = if let Some(change_address) = self.change_address.take() {
+                change_address
+            } else if !self.spends.is_empty() {
+                (
+                    self.spends[0].extsk.expsk.ovk,
+                    PaymentAddress {
+                        diversifier: self.spends[0].diversifier,
+                        pk_d: self.spends[0].note.pk_d.clone(),
+                    },
+                )
+            } else {
+                return Err(Error(ErrorKind::NoChangeAddress));
+            };
+
+            self.add_sapling_output(change_address.0, change_address.1, Amount(change), None)?;
+        }
```

Comment:
We could at this point, in theory, check (or assert?) that valuebalance is zero correct?

Reply:
No; `value_balance` should equal the fee at this point.